Sample Course now has various autograding/TA grading assignments

holzbh opened the pull request:
Addresses most of #830. A new issue (#1164) has been opened for creating assignments with various late day usages.

MasterOdin commented on .setup/bin/setup_sample_courses.py:
```diff
@@ -443,6 +446,185 @@ def create_gradeable_submission(src, dst):
         os.remove(zip_dst)
 
 
+def make_sample_json():
```
So we're not always guaranteed to be building the sample course (if we only pass in say tutorial or whatever), I'd rather move the "make_sample_json()" function inside the course loop (right after the create function?) and make it a configuration option in the yml file (false by default).

holzbh replied:
Can do, makes sense to me.

holzbh commented on .setup/bin/setup_sample_courses.py:
```diff
@@ -443,6 +446,185 @@ def create_gradeable_submission(src, dst):
+        gradeables_percentages[-1] += gradeable_percentage_left
+
+    # Compute totals and write out each syllabus bucket in the "gradeables" field of customization.json
+    # TODO: Figure out what to do about "none" bucket items - does Rainbow Grades support this?
```
It's a real bucket yes, but I don't think I've seen people type in

{
type: none,
count: 3,
percent 0.25,
[{id: .... }]
}

The syllabus bucket in the DB previously had nothing to do with how rainbow grades handled the type, but going forward as we offer a tool for generating customization files automatically, we would like to start using the syllabus buckets.

MasterOdin replied:
Okay.

holzbh replied:
Reading over this, I think you were misinterpreting my comment. I wasn't considering what to do if there is no bucket type, but more can Rainbow Grades handle "None (for practice only)".

As it turns out, Rainbow Grades [https://github.com/Submitty/Submitty/blob/master/RainbowGrades/main.cpp#L283-L298](will) look for anything starting with "none" at the moment (as well as a few other parsing oddities). So I can remove the comment.

holzbh commented on .setup/bin/setup_sample_courses.py:
```diff
@@ -443,6 +446,185 @@ def create_gradeable_submission(src, dst):
+    # TODO: Figure out what to do about "none" bucket items - does Rainbow Grades support this?
+    bucket_no = 0
+    for bucket,g_list in gradeables.items():
+        gradeables_json_output += "    {\n" +\
```
Right, forgot about just dumping using json.dump() will adapt everywhere I'm currently writing out the whole mess by hand.

holzbh commented on .setup/bin/setup_sample_courses.py:
```diff
@@ -489,7 +671,7 @@ def __init__(self, user):
             self.rotating_section = int(user['rotating_section'])
         if 'grading_registration_section' in user:
-            self.grading_registration_section = int(user['grading_registration_section'])
+            self.grading_registration_section = user['grading_registration_section']
```
I added support for multiple grading registration sectiosn. By having the typecast we were limiting TAs to a max of 1 section per course, but this is too inflexible.

MasterOdin replied:
Okay, yeah, you do the typecast later when you actually use this field. Got it.

holzbh commented on tests/rainbowGrades/test_sample.py:
```diff
@@ -122,7 +122,7 @@ def sample_rainbow_grades_test():
                 elif len(line) >= 18 and line[:18] == "REPORTS_DIRECTORY=":
                     make_file.write(os.path.join("REPORTS_DIRECTORY=__INSTALL__FILLIN__SUBMITTY_DATA_DIR__", "courses",
-                                                 "s17", "sample", "reports") + "\n")
+                                                 "f17", "sample", "reports") + "\n")
```
Can do.

holzbh commented:
* Did the changes requested in the reviews

* Moved the JSON generation to be a method of Courses ( `make_course_json()` ) and made it work for any course.

* Extended courses.yml to support a flag for generation of customization.json per course

* Fixed some error reporting in test_sample.py, moved it to use the generated customization_sample.json

* Updated the expected Rainbow Grades output in the test to reflect TA names being used and course name being used.
